### PR TITLE
MFLT-8016 Document dev-mode and new commands for Linux SDK 1.2.0

### PR DESCRIPTION
**Pull request by sarfata**
MFLT-8016 Document dev-mode and new commands for Linux SDK 1.2.0

  ### Summary

This documents the new SDK features included in Linux SDK 1.2.0.

  - memfaultctl and all it's new features

I have also split the getting started into:

  - A quickstart guide with the instructions to use the Docker image (this used
  to be only in a README in the meta-memfault-example layer) but I think it's
  very valuable and wanted to make this more visible.

  - An integration guide (used to be called getting started guide). This matches
  what we do on MCUs.

And I have renamed some pages to avoid having linux multiple times in the page
URL (again this matches what we do for MCUs and most Android pages).

For each renamed page I have added a redirect in the Netlify configuration.

  ### Test Plan

- Make sure the doc builds.
- Make sure all renamed pages have an entry in the `_redirects` file.
- Proof read for grammar and spelling.
- Make sure all links have been updated to the new URLs

### CODE REVIEW

**Comment by sarfata** [comment]
Current dependencies on/for this PR:
* main
  * **PR #350** <a href="https://app.graphite.dev/github/pr/memfault/memfault-docs/350" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/memfault/memfault-docs/350?utm_source=stack-comment).

File: docs/linux/ota.mdx [hunk 1]
```diff
@@ -73,8 +73,19 @@ page in the Memfault UI and copy/paste the token from the "Project Key" section.
...
-[memfault cli tool](https://docs.memfault.com/docs/ci/install-memfault-cli). In
-this tutorial we will make use of the CLI client which can be installed via pip:
+[memfault cli tool](https://docs.memfault.com/docs/ci/install-memfault-cli).
+
+:::tip
+
+Memfault CLI tool (`memfault-cli`) is a Python utility that is meant to be
+installed on development machines and CI infrastructure. It interacts with
+Memfault API. It is different from the much smaller (and binary) `memfaultctl`
+which interacts with `memfaultd`.
```
Comment (martijnthe) [inline docs/linux/ota.mdx:83]: Maybe split in two sentences?

```suggestion
Memfault API. It is different from `memfaultctl`, which interacts with `memfaultd`. As a small, native binary, it is optimized to be run on embedded devices.
```

File: docs/linux/quickstart.mdx [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+[integration-guide]: /docs/linux/integration-guide/
+
+## Build and Run Linux with Memfault SDK
+
+### Create a Memfault Project
+
+Go to [app.memfault.com](https://app.memfault.com) and from the "Select a
+Project" dropdown, click on "Create Project". Once you're done, you can find a
+project key, referenced as YOUR_PROJECT_KEY in this document, in the Project
```
Comment (martijnthe) [inline docs/linux/quickstart.mdx:22]:
```suggestion
project key, referenced as `YOUR_PROJECT_KEY` in this document, in the Project
```

File: docs/linux/quickstart.mdx [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+This project is configured to build an image for the `arm64` architecture on a
+QEMU (emulated) machine.
+
+To build the image, run:
+
+```
+$ bitbake memfault-image
+```
+
+This will take around two hours the first time. Yocto will cache compilation
```
Comment (martijnthe) [inline docs/linux/quickstart.mdx:55]: Depends a lot on your machine... for me it's been around 30 mins or so.

```suggestion
This will take a long time the first time. Yocto will cache compilation
```

File: docs/linux/quickstart.mdx [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+results and future builds will be much faster. We use named Docker volumes to
+persist source code and build artifacts across reboots.
```
Comment (martijnthe) [inline docs/linux/quickstart.mdx:57]:
```suggestion
persist source code and build artifacts beyond the container lifetime.
```

File: docs/linux/quickstart.mdx [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+### Build your Yocto-based image
+
+This project is configured to build an image for the `arm64` architecture on a
+QEMU (emulated) machine.
+
+To build the image, run:
+
+```
+$ bitbake memfault-image
```
Comment (martijnthe) [inline docs/linux/quickstart.mdx:52]: We've got the `b` shortcut for this, but maybe it's less magical this way.

File: docs/linux/quickstart.mdx [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+[dockerfile]:
+  https://github.com/memfault/memfault-linux-sdk/blob/-/docker/Dockerfile
+[run.sh]: https://github.com/memfault/memfault-linux-sdk/blob/-/docker/run.sh
+
+### Run the image with QEMU
+
+To start QEMU and run the image you have just build, we provide a convenient
+shortcut:
+
```
Comment (martijnthe) [inline docs/linux/quickstart.mdx:75]: Maybe mention that `q` is more or less the same as what `runqemu qemuarm64 slirp nographic` in newer Yocto releases runs? People familiar with `runqemu` might be wondering why we're not using it. 

 We don't use it because `runqemu` was not available in older Yocto releases that we support (currently only Dunfell).

File: docs/linux/reboot-reason-tracking.mdx [hunk 7]
```diff
@@ -320,14 +332,30 @@ the topics discussed together.
   {/* source: https://whimsical.com/linux-sdk-reboot-reasons-flowchart-4AoG7cAzbC1DMCw5F9w2X3 */}
 </ImageFigure>
 
-[docs-ota]: /docs/linux/linux-releases-integration-guide
+## Test your integration
+
+You can easily test your integration with the `memfaultctl reboot <reason>`
+command. It accepts an integer reboot reason (defaults to 0 `UNKNOWN_REASON`),
```
Comment (martijnthe) [inline docs/linux/reboot-reason-tracking.mdx:338]: Should we refer/link to the `docs-reboot-reason-ids` here? I know it's mentioned 2 sections up from here, but it's easy to miss / doesn't hurt to repeat here?

File: docs/linux/reference-memfaultctl-cli.mdx [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+- `write-attributes` events will be flushed immediately to Memfault ;
+
+To enable developer mode, use `memfaultctl enable-dev-mode`. This setting will
+be persisted by `memfaultd` accross reboots.
+
+## `memfaultctl` commands
+
+### `enable-data-collection`
+
+Call this command when your user has accepted the collection of usage and
```
Comment (martijnthe) [inline docs/linux/reference-memfaultctl-cli.mdx:58]:
```suggestion
Call this command when your user has given consent to the collection of usage and
```

File: docs/linux/reference-memfaultctl-cli.mdx [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+### `show-settings`
+
+Prints the current configuration of `memfaultd`.
+
+Example:
+
+```console
+# memfaultctl show-settings
+Base configuration (/etc/memfaultd.conf):
+  { "queue_size_kib": 1024, "data_dir": "\/media\/memfault", "refresh_interval_seconds": 3600, "enable_data_collection": false, "enable_dev_mode": false, "software_version": "0.0.1", "software_type": "main", "project_key": "INSERT_PROJECT_KEY_HERE", "base_url": "https:\/\/device.memfault.com", "swupdate_plugin": { "input_file": "\/etc\/swupdate.cfg", "output_file": "\/tmp\/swupdate.cfg" }, "reboot_plugin": { "last_reboot_reason_file": "\/media\/last_reboot_reason", "uboot_fw_env_file": "\/etc\/fw_env.config" }, "collectd_plugin": { "header_include_output_file": "\/tmp\/collectd-header-include.conf", "footer_include_output_file": "\/tmp\/collectd-footer-include.conf", "non_memfaultd_chain": "", "write_http_buffer_size_kib": 64, "interval_seconds": 3600 }, "coredump_plugin": { "coredump_max_size_kib": 96000, "rate_limit_count": 5, "rate_limit_duration_seconds": 3600, "storage_min_headroom_kib": 10240, "storage_max_usage_kib": 0 } }
```
Comment (martijnthe) [inline docs/linux/reference-memfaultctl-cli.mdx:101]: OT: wish we pretty-printed this.. 8/

File: docs/linux/reference-memfaultctl-cli.mdx [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+  coredump
+
+```
+
+### `sync`
+
+Immediately process `memfaultd` queue.
+
+`memfaultd` stores events in a queue and only processes it at a fixed interval
+(by default 1 hour). Use `memfaultctl sync` to force `memfaultd` to immediately
```
Comment (martijnthe) [inline docs/linux/reference-memfaultctl-cli.mdx:127]: Mention the config key that controls the interval + link to the reference?

File: docs/linux/reference-memfaultctl-cli.mdx [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+
+Use this command to create or update [device attributes][attributes]. You can
+provide multiple attributes in one call.
+
+```console
+# memfaultctl write-attributes APP_VERSION=1.4.2 ACTIVATED=true
+```
+
+Attributes uploaded with this command cannot be used as metrics in the Memfault
+panel. Refer to our [metrics][docs-metrics] guide to upload application-specific
```
Comment (martijnthe) [inline docs/linux/reference-memfaultctl-cli.mdx:153]:
```suggestion
Attributes uploaded with this command can only be used as device attributes and not as time series metrics in Memfault. Refer to our [metrics][docs-metrics] guide to upload application-specific
```

File: docs/linux/reference-memfaultctl-cli.mdx [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+metrics.
+
+The Memfault SDK does not attempt to avoid unnecessary calls. Every call to this
+command will eventually trigger a network connection. You should avoid setting
```
Comment (martijnthe) [inline docs/linux/reference-memfaultctl-cli.mdx:157]:
```suggestion
command will eventually trigger a network request. You should avoid setting
```

File: docs/linux/reference-memfaultctl-cli.mdx [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+the same value repeatedly as this would increase network usage.
+
+[docs-metrics]: /docs/linux/metrics#application-metrics
+
+#### Attribute types
+
+`memfaultctl` will try to guess the type of the attributes:
+
+- `true`/`false` are converted to boolean values ;
+- Numbers are converted to float values ;
```
Comment (martijnthe) [inline docs/linux/reference-memfaultctl-cli.mdx:167]:
```suggestion
- `true`/`false` are converted to boolean values;
- Numbers are converted to float values;
```

File: docs/linux/reference-memfaultd-cli.mdx [hunk 14]
```diff
@@ -28,35 +29,19 @@ Usage: memfaultd [OPTION]...
 The `--config-file` path defaults to `/etc/memfaultd.conf`. The settings you add
 in `/etc/memfaultd.conf` extend the built-in configuration file.
 
-[source-builtin-conf]:
-  https://github.com/memfault/memfault-linux-sdk/blob/-/meta-memfault/recipes-memfault/memfaultd/files/memfaultd/builtin.conf
+:::tip
 
-#### Sample output from `memfaultd --show-settings`
+We recommend using [`memfaultctl`][docs-memfaultctl] to enable/disable data
+collection and view current settings.
```
Comment (martijnthe) [inline docs/linux/reference-memfaultd-cli.mdx:35]: Maybe mention the version from which this is true?

... as of v1.2.0

File: docs/linux/introduction.mdx [hunk 15]
```diff
@@ -10,15 +10,16 @@ your fleet.
...
+well-established, battle-tested open-source software. The daemon `memfaultd`
+implements Memfault-specific features and also acts as a configuration agent.
+`memfaultctl` is a command line utility to adjust `memfaultd` configuration and
+trigger specific events for testing purposes.
 
 <p>
   <figure>
     <img
-      src="/img/docs/linux/sdk-overview-1.0.0.svg"
+      src="/img/docs/linux/sdk-overview-1.2.0.svg"
```
Comment (martijnthe) [inline docs/linux/introduction.mdx:22]: Digging the new `memfaultctl` mention in the diagram!

**Review by martijnthe — APPROVED** [review]

File: docs/linux/quickstart.mdx (hunk 6 above)
Comment (sarfata) [reply docs/linux/quickstart.mdx:75]: @martijnthe 

I have added:
```
:::tip

The `q` shortcut executes our own [`runqemu.py`][runqemu] which is very similar
to `runqemu qemuarm64 slirp nographic` in newer Yocto distributions.

We maintain compatibility with older versions of Yocto in which `runqemu` did not support the dual-copy SWUpdate setup that we use here. This is why we provide our own.

:::
```

I just want to make sure that this is true. [Source](https://github.com/memfault/memfault-linux-sdk-internal/blob/master/DEVELOPMENT.md?plain=1#L165)

File: docs/linux/reboot-reason-tracking.mdx (hunk 7 above)
Comment (sarfata) [reply docs/linux/reboot-reason-tracking.mdx:338]: agreed

File: docs/linux/reference-memfaultctl-cli.mdx (hunk 9 above)
Comment (sarfata) [reply docs/linux/reference-memfaultctl-cli.mdx:101]: Yeah but it would be very long. As we add more config options I think we can stop dumping the config file.

**Review by fnune — APPROVED** [review]

File: static/_redirects [hunk 16]
```diff
@@ -17,3 +17,10 @@
 /docs/platform/device-attributes /docs/platform/metrics#device-attributes
```
Comment (martijnthe) [inline static/_redirects:17]: (Comment about the .svg)

Same comment as in the other PR -- the arrows to Memfault Data & Config are placed oddly. Add a curve?

File: docs/linux/reference-memfaultctl-cli.mdx (hunk 9 above)
Comment (martijnthe) [reply docs/linux/reference-memfaultctl-cli.mdx:101]: For our own debugging it would be handy still. Maybe we can hide it behind a `--verbose` option at some point.

File: docs/linux/reference-memfaultctl-cli.mdx [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+  write-attributes <VAR1=VAL1 ...>           : Write device attribute(s) to Memfaultd
+
+```
+
+The `--config-file` path defaults to `/etc/memfaultd.conf`. The settings you add
+in `/etc/memfaultd.conf` extend the built-in configuration file.
+
+## Developer mode
+
+Enabling developer mode will change some settings in `memfaultd` to make
```
Comment (martijnthe) [inline docs/linux/reference-memfaultctl-cli.mdx:44]: nit
```suggestion
Enabling developer mode will change some settings of `memfaultd` to make
```

File: docs/linux/reference-memfaultctl-cli.mdx [hunk 18]
```diff
@@ -0,0 +1,194 @@
...
+development and testing easier:
+
+- Coredumps will not be rate-limited on device. All crashes will be reported to
+  Memfault immediately (queue is processed immediately);
+- `write-attributes` events will be flushed immediately to Memfault.
+
+To enable developer mode, use `memfaultctl enable-dev-mode`. This setting will
+be persisted by `memfaultd` accross reboots.
```
Comment (martijnthe) [inline docs/linux/reference-memfaultctl-cli.mdx:52]: `memfaultctl` is actually writing the config file to disk. I don't think it's very relevant what does it, so I would drop this.

Same thing below a few more times.

```suggestion
be persisted accross reboots.
```

File: docs/linux/reboot-reason-tracking.mdx [hunk 19]
```diff
@@ -237,6 +237,18 @@ file:
...
+
+To save a device-specific reboot reason and immediately reboot you can use
+`memfaultctl`.
+
+```console
+# memfaultctl reboot --reason <reboot reason integer>
+```
+
+This command simply writes the reboot reason in the `last_reboot_reason_file`
+and then runs the `reboot` command.
```
[the remark this answers is not in the document]
Comment (martijnthe) [reply docs/linux/reboot-reason-tracking.mdx:250]: iirc, `reboot` is just a symlink to systemctl on systemd systems!
multi-call binary again ;) (so much for "Make each program do one thing well")

File: docs/linux/quickstart.mdx (hunk 6 above)
Comment (martijnthe) [reply docs/linux/quickstart.mdx:75]: Yup, 100% accurate!

**Review by martijnthe — APPROVED** [review]

File: docs/linux/reboot-reason-tracking.mdx (hunk 19 above)
[the remark this answers is not in the document]
Comment (fnune) [reply docs/linux/reboot-reason-tracking.mdx:250]: image

File: docs/linux/reboot-reason-tracking.mdx [hunk 20]
```diff
@@ -237,6 +237,18 @@ file:
   during shutdown. In the flowchart below, the green boxes indicate when this
   code may be run (pick one).
 
+#### Rebooting with a reason
+
+To save a device-specific reboot reason and immediately reboot you can use
+`memfaultctl`.
+
```
[the remark this answers is not in the document]
Comment (sarfata) [reply docs/linux/reboot-reason-tracking.mdx:244]: Added a link to the memfaultctl reference page.

File: static/_redirects (hunk 16 above)
Comment (sarfata) [reply static/_redirects:17]: Good catch!

File: static/_redirects [hunk 21]
```diff
@@ -17,3 +17,10 @@
...
 /docs/mcu/self-serve/* /docs/mcu/embedded-self-serve/:splat
 /docs/best_practices/* /docs/best-practices/:splat
+
+# Renamed some pages for the Linux SDK
+/docs/linux/linux-getting-started-guide /docs/linux/integration-guide
+/docs/linux/linux-releases-integration-guide /docs/linux/ota
+/docs/linux/linux-coredumps /docs/linux/coredumps
+/docs/linux/linux-metrics /docs/linux/metrics
+/docs/linux/linux-memfault-hawkbit-comparison /docs/linux/memfault-hawkbit-comparison
```
[the remark this answers is not in the document]
Comment (sarfata) [reply static/_redirects:26]: Thanks for pointing this out => [MFLT-8328](https://memfault.myjetbrains.com/youtrack/issue/MFLT-8328) Update linux doc links on mftl.io to be up to date

I looked for other redirects mechanisms but that `_redirects` file seems to be what we are using now and also what Docusaurus recommends (they have no other mechanism built-in, there is a client side redirect plugin but they recommend using server side redirects from the provider). I guess if we move to something else we will probably migrate what we have in the `_redirects` file.

File: docs/linux/quickstart.mdx [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+### Inspect the integration
+
+Login as `root` and make sure to enable data-collection (disabled by default).
+
+```
+# memfaultctl enable-data-collection
+```
+
+You can now use `memfaultctl` to generate events on the device. They will appear
+in your Memfault Dashboard.
```
[the remark this answers is not in the document]
Comment (fnune) [reply docs/linux/quickstart.mdx]: Here's the article I usually go by when choosing whether to pick `target="_blank"` or not: https://css-tricks.com/use-target_blank/

tl;dr: the only good reason to break from the default is when the user is working on something and they would lose that state if they left the page.

In this case, if the user left the page, they wouldn't lose any state: not even their scroll position, which is handled correctly in the docs.

An argument not mentioned in the article is that `target="_blank"` leaves the user without a choice of where their link will open, but the default honors this (the user can always Ctrl+Click mousewheel-click).

[the remark this answers is not in the document]
Comment (fnune) [reply docs/linux/quickstart.mdx]: These are the reasons why I chose to not use `target="_blank"` as a default, but also all good reasons not to use it in this case either.
